ENH: Add conditional to adapt behaviour depending on astropy version and warn user

vandalt opened the pull request:
This is a follow-up for #60 by @neutrinoceros. As discussed there, I added a warning that commentary cards are removed. I also changed `working version` to `5.0rc1`, the first release candidate with the upstream fix included. I also added tests to 1) check that the card is included in 5.0 and 2) Check that the warning is issued before 5.0.

neutrinoceros reviewed (changes requested):
Looks good, thank you for continuing this work !
I made a couple suggestions to improve the tests, but in general I'm in favour of this simply superseding my own PR, so I'll close #60

neutrinoceros commented:
Another important thing is that we should test this properly using the pre release. I'll open a PR to do just that

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -66,13 +66,14 @@ def test_add_infos_header_commentary(commentary_infos):
-    if Version(astropy_version) < Version(working_version):
-        assert "HISTORY" not in commentary_infos.hdr
+ASTROPY_VERSION = Version(astropy.__version__)
+@pytest.mark.skipif(ASTROPY_VERSION < Version("5.0rc"), reason="...")
```
lol. I meant `"..."` to be a placeholder, please be less lazy than I :)

vandalt replied:
Oops haha, I'll change that locally in a few minutes :)

vandalt commented:
Thanks @neutrinoceros ! For the first suggestion, I kept the two tests for now. I'm not opposed to removing the "drops" one. I have little experience with unit tests, so I'd let you decide what to do here. My initial idea was to test the behaviour because there is a warning related to it, so if the behaviour is fixed, a failing test would flag that the warning needs to be removed.

vandalt commented on amical/tests/test_extraction.py:
```diff
@@ -1,42 +1,102 @@
+    return hdr
+
+
+@pytest.fixture()
+def astropy_versions():
+
+    astropy_version = astropy.__version__
+    working_version = "5.0rc1"
+
+    return astropy_version, working_version
```
Using a constant for this seems better. Should we also do it for "```ASTROPY_WORKING```" ?

vandalt replied:
Fixed in a separate commit (to also remove any reference to `astropy_versions` at the same time)

neutrinoceros commented:
> My initial idea was to test the behaviour because there is a warning related to it, so if the behaviour is fixed, a failing test would flag that the warning needs to be removed.

It's a good idea. In general, it's good to test failing cases, not just successes. Another rule of thumb is to avoid testing implementation details instead of behaviour, because it tends to rigidify the design, and it's a burden to maintain.

So ideally here we should have:
- a test that checks that the warning is raised if astropy < 5.0 (using `@pytest.mark.skipif`)
- another test, identical to the one we dedicate to astropy 5.0, but reserved for older versions (using `@pytest.mark.skipif` again) AND decorated with `@pytest.mark.xfail` (see https://docs.pytest.org/en/6.2.x/skipping.html#xfail-mark-test-functions-as-expected-to-fail)

vandalt commented on amical/mf_pipeline/bispect.py:
```diff
@@ -977,27 +979,38 @@ def _compute_phs_error(complex_bs, fitmat, index_mask, npix, imsize=3):
+    working_version = "5.0rc1"
+    if Version(astropy_version) < Version(working_version):
+        if verbose:
+            cprint(
```
Done! I also moved things around to warn only if there are commentary cards in the header, to avoid warning when nothing is removed from the header.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -7,6 +7,11 @@
 
+# Astropy versions for
+ASTROPY_VERSION = Version(astropy.__version__)
+ASTROPY_WORKING = Version("5.0rc")
```
I'd suggest not using a global const for this one. For now we only have one place where we adapt behaviour as a function of satrapy's version, but in the future there could be more, in which case this variable name will become unclear and confusing.

vandalt replied:
Fixed by the last commit. 5.0rc is used everywhere directly now.

neutrinoceros commented on amical/mf_pipeline/bispect.py:
```diff
@@ -989,13 +990,16 @@ def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix, verbose=True
-    if Version(astropy_version) < Version(working_version):
+    hdr_commentary_keys = fits.Card._commentary_keywords
+    if any(hck in hdr for hck in hdr_commentary_keys) and (
+        Version(astropy_version) < Version(working_version)
```
same as my previous comment; I'd suggest storing the runtime version in a const (`ASTROPY_VERSION = Version(astropy.__version__)`) but not the requirement, which should be parsed only if needed, e.g. directly in the `if` statement

vandalt replied:
Fixed by the last commit. 5.0rc is used directly in the `if` statement now.

neutrinoceros commented on amical/mf_pipeline/bispect.py:
```diff
@@ -989,13 +990,16 @@ def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix, verbose=True
                 f" version < {working_version}. Your astropy version is"
                 f" {astropy_version}",
-                "green",
+                RuntimeWarning,
```
I researched it (I can never remember what warning types to use), but this seems to be the best choice to me, well done !

vandalt replied:
I also never remember Python warnings, I agree `RuntimeWarning` seems like the best option, but I'm open to suggestions if anyone else has a better idea! I'll mark as resolved for now.

neutrinoceros commented on amical/mf_pipeline/bispect.py:
```diff
@@ -977,27 +980,40 @@ def _compute_phs_error(complex_bs, fitmat, index_mask, npix, imsize=3):
 
 
-def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix):
+def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix, verbose=True):
```
I don't think we should support warning deactivation, especially via boolean arguments (see for instance https://ariya.io/2011/08/hall-of-api-shame-boolean-trap)
```suggestion
def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix):
```

vandalt replied:
I agree! I fixed this in a separate commit to remove the related `if` statement at the same time. Marking as resolved.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -1,42 +1,95 @@
+    assert "HISTORY" in commentary_infos.hdr
+
+
+def test_astropy_version_warning(infos, commentary_hdr):
```
This still needs a `pytest.mark.skipif` decorator, or the test will fail on astropy 5.0

vandalt replied:
Yes sorry I was still working on that locally when I pused the other changes. It should have been uploaded now.

neutrinoceros commented on amical/mf_pipeline/bispect.py:
```diff
@@ -41,6 +41,9 @@
 from amical.tools import cov2cor
 
 
+ASTROPY_VERSION = astropy.__version__
```
sorry to be such a pain but...
```suggestion
ASTROPY_VERSION = Version(astropy.__version__)
```

vandalt replied:
Fixed in a separate commit to remove the other (now redundant) `Version()` in the if statement.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -1,42 +1,116 @@
+ASTROPY_VERSION = Version(astropy.__version__)
+
+
+@pytest.fixture()
+def infos():
+    # SimulatedData avoids requiring extra keys in infos
+    return munch.Munch(orig="SimulatedData", instrument="unknown")
+
+
+@pytest.fixture()
+def commentary_hdr():
+    # Create a fits header with commentary card
+    hdr = fits.Header()
+    hdr["HISTORY"] = "History is a commentary card"
+    return hdr
```
btw there is no reason to use fixtures for such small data bits. Here's another advice for writing tests: while it's great to follow the DRY principles for production code, it's actually okay to repeat -small amounts of- code in tests
Using fixtures has a cost on readability/maintainability.
It is worth using them when setting up and tearing down a context is costly, or when it becomes clear that large portions of test code are repeated verbatim. It's not the case here.

Note for disclosure that this is my _opinion_, I don't claim to be a test expert, but I _have_ fallen into common pitfalls in the past so I'm trying to avoid them here.

vandalt replied:
Good to know! Thanks :) I removed both. I kept it for `commentary_infos`, as it is slightly longer, but I can change that as well if needed, just let me know.

neutrinoceros commented:
we're getting there, thank you for your patience @vandalt !

vandalt commented:
Thanks for your detailed comments @neutrinoceros ! I think I addressed all of them, but let me know if I forgot somethign or if there are others.

neutrinoceros commented on amical/mf_pipeline/bispect.py:
```diff
@@ -997,7 +1010,6 @@ def _add_infos_header(infos, hdr, mf, pa, filename, maskname, npix):
         # https://github.com/astropy/astropy/issues/11866
         # Resolved upstream with
         # https://github.com/astropy/astropy/pull/11923
-        hdr_commentary_keys = fits.Card._commentary_keywords
```
I'm not sure why this line was removed. @vandalt could you explain your reasoning ?

vandalt replied:
It was moved out of the `if` statement to issue the warning only when there is actually a "commentary key" in the header (i.e. when something will indeed be removed form the header). I can add a comment before the `if` to clarify that.

vandalt replied:
I added the comment, marking as resolved.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -22,21 +45,76 @@ def test_add_infos_simulated():
     assert infos["date-obs"] == hdr["DATE-OBS"]
     assert infos["telescop"] == hdr["TELESCOP"]
     assert "observer" not in infos  # Keys that are not in hdr should not be in infos
+    assert "observer" not in infos.hdr  # Keys that are not in hdr should still not be
```
looks like the comment was truncated ?

vandalt replied:
Indeed, I will clarify this in a new commit.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -22,21 +45,76 @@ def test_add_infos_simulated():
 
 
-def test_add_infos_header_commentary():
+@pytest.mark.filterwarnings("ignore: Commentary cards")
```
Let's make this as specific as possible
```suggestion
@pytest.mark.filterwarnings("ignore: Commentary cards are removed from the header")
```

vandalt replied:
Do you want me to use the full warning message or is this new version enough?

vandalt replied:
I merged the change, but I'll keep it "unresolved" just in case I need to add the full message.

neutrinoceros replied:
well I don't think we can/should use the full warning because it includes data determined at runtime, which is what I meant by "as specific as possible"

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -22,21 +45,76 @@ def test_add_infos_simulated():
+    assert "HISTORY" in commentary_infos.hdr
+
+
+@pytest.mark.skipif(
+    ASTROPY_VERSION >= Version("5.0rc"),
+    reason="Munch can handle commentary cards for Astropy 5.0+",
+)
```
Actually there's no reason to skip this, I think ? marking it as conditionally `xfail` is all we need
```suggestion
```

vandalt replied:
Solved by last commit using xfail only and removing redundant tests.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -22,21 +45,76 @@ def test_add_infos_simulated():
+    assert "HISTORY" in commentary_infos.hdr
+
+
+@pytest.mark.skipif(
```
Same reasoning as above :)
```suggestion
@pytest.mark.xfail(
```

vandalt replied:
Solved by last commit using xfail only and removing redundant tests.

vandalt replied:
Solved by last commit using xfail only and removing redundant tests.

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -22,21 +45,76 @@ def test_add_infos_simulated():
+        infos = _add_infos_header(infos, hdr, mf, 1.0, "afilename", "amaskname", 1)
+
+
+@pytest.mark.skipif(
+    ASTROPY_VERSION < Version("5.0rc"),
+    reason="Astropy < 5.0 should raise a warning for commentary cards",
+)
```
```suggestion
```

neutrinoceros reviewed (changes requested):
I think this should be my final round of review. I apologies for misleading you into combining `xfail` and `skipif`, I realised while reading this that we should probably never have to use more than one.

neutrinoceros commented:
It'd be useful to run the "bleeding edge" job I just merged into master here, to check that your tests behave correctly with astropy's dev version.
Basically you'd need to rebase against the current state of the master branch and run the job manually on your fork. If you need help for that don't hesitate to ask

vandalt commented:
I rebased on master and also rebased the master of my fork on master. I'm not sure how to run the "bleeding edge" CI though. I did run pytest in a virtualenv with the `5.0rc1` version of astropy and all tests passed.

neutrinoceros commented:
On your fork, go to the "Actions" section, select the job as shown, and click the "run workflow" you can see on the right (select the branch attached to this PR)
![Screenshot 2021-11-03 at 16 19 14](https://user-images.githubusercontent.com/14075922/140089299-e4e732de-c4f5-49c4-a0c6-b755b5292b78.png)

neutrinoceros commented:
No you don't. Actually you could deactivate actions on your fork in general, it's not necessary.

neutrinoceros commented:
so the job is here https://github.com/vandalt/AMICAL/runs/4094406054?check_suite_focus=true
it WILL fail because of Codecov, I just want to make sure it doesn't fail for a different reason

vandalt commented:
Indeed it did, but the tests worked as expected I think.
![image](https://user-images.githubusercontent.com/35547958/140091888-89bfc9e0-4986-423c-9bec-f3aef371a031.png)
Or at least it gives the same result as my local virtualenv with astropy 5.

neutrinoceros commented:
I confirm that the results are consistent with running locally with the release candidate. Hurray !
Now for one ultimate, and probably quick, round

neutrinoceros commented on amical/tests/test_extraction.py:
```diff
@@ -12,31 +35,53 @@ def test_add_infos_simulated():
+    reason="Munch cannot handle commentary cards for Astropy < 5.0",
+)
+def test_commentary_infos_keep(commentary_infos):
+    assert "HISTORY" in commentary_infos.hdr
```
I do not understand why Codecov reports this as not hit. It runs on my machine whatever version of astropy I use. Any clue ? (not needed to get this merged)

vandalt replied:
No sorry... Works on my machine as well, and I don't know much about Codecov

vandalt commented:
Thanks for the help on this @neutrinoceros. Sorry I was away today, so your last suggestion was accepted via mobile and the commit message is not very informative.
